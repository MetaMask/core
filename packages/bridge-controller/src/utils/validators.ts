import { isValidHexAddress } from '@metamask/controller-utils';
import {
  string,
  boolean,
  number,
  type,
  is,
  record,
  array,
  nullable,
  optional,
  enums,
  define,
  union,
} from '@metamask/superstruct';
import { isStrictHexString } from '@metamask/utils';

import type { BridgeAsset, FeatureFlagResponse, QuoteResponse } from '../types';
import { ActionTypes, BridgeFlag, FeeType } from '../types';

const HexAddressSchema = define('HexAddress', (v: unknown) =>
  isValidHexAddress(v as string, { allowNonPrefixed: false }),
);

const HexStringSchema = define('HexString', (v: unknown) =>
  isStrictHexString(v as string),
);

export const truthyString = (s: string) => Boolean(s?.length);
const TruthyDigitStringSchema = define(
  'TruthyDigitString',
  (v: unknown) =>
    truthyString(v as string) && Boolean((v as string).match(/^\d+$/u)),
);

const ChainIdSchema = number();

const BridgeAssetSchema = type({
  chainId: ChainIdSchema,
  address: string(),
  assetId: string(),
  symbol: string(),
  name: string(),
  decimals: number(),
  icon: optional(string()),
  iconUrl: optional(string()),
});

export const validateFeatureFlagsResponse = (
  data: unknown,
): data is FeatureFlagResponse => {
  const ChainConfigurationSchema = type({
    isActiveSrc: boolean(),
    isActiveDest: boolean(),
    refreshRate: optional(number()),
    topAssets: optional(array(string())),
  });

  const ConfigSchema = type({
    refreshRate: number(),
    maxRefreshCount: number(),
    support: boolean(),
    chains: record(string(), ChainConfigurationSchema),
  });

  // Create schema for FeatureFlagResponse
  const FeatureFlagResponseSchema = type({
    [BridgeFlag.EXTENSION_CONFIG]: ConfigSchema,
    [BridgeFlag.MOBILE_CONFIG]: ConfigSchema,
  });

  return is(data, FeatureFlagResponseSchema);
};

export const validateSwapsTokenObject = (
  data: unknown,
): data is BridgeAsset => {
  return is(data, BridgeAssetSchema);
};

export const validateQuoteResponse = (data: unknown): data is QuoteResponse => {
  const FeeDataSchema = type({
    amount: TruthyDigitStringSchema,
    asset: BridgeAssetSchema,
  });

  const ProtocolSchema = type({
    name: string(),
    displayName: optional(string()),
    icon: optional(string()),
  });

  const StepSchema = type({
    action: enums(Object.values(ActionTypes)),
    srcChainId: ChainIdSchema,
    destChainId: optional(ChainIdSchema),
    srcAsset: BridgeAssetSchema,
    destAsset: BridgeAssetSchema,
    srcAmount: string(),
    destAmount: string(),
    protocol: ProtocolSchema,
  });

  const RefuelDataSchema = StepSchema;

  const QuoteSchema = type({
    requestId: string(),
    srcChainId: ChainIdSchema,
    srcAsset: BridgeAssetSchema,
    srcTokenAmount: string(),
    destChainId: ChainIdSchema,
    destAsset: BridgeAssetSchema,
    destTokenAmount: string(),
    feeData: record(enums(Object.values(FeeType)), FeeDataSchema),
    bridgeId: string(),
    bridges: array(string()),
    steps: array(StepSchema),
    refuel: optional(RefuelDataSchema),
  });

  const TxDataSchema = type({
    chainId: number(),
    to: HexAddressSchema,
    from: HexAddressSchema,
    value: HexStringSchema,
    data: HexStringSchema,
    gasLimit: nullable(number()),
  });

  const QuoteResponseSchema = type({
    quote: QuoteSchema,
    approval: optional(TxDataSchema),
    trade: union([TxDataSchema, string()]),
    estimatedProcessingTimeInSeconds: number(),
  });

  return is(data, QuoteResponseSchema);
};
