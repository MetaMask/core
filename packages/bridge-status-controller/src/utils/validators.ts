import {
  object,
  string,
  boolean,
  number,
  optional,
  enums,
  union,
  type,
  nullable,
  assert,
} from '@metamask/superstruct';

import { StatusTypes } from '../types';

export const validateBridgeStatusResponse = (data: unknown) => {
  const ChainIdSchema = union([number(), string()]);

  const AssetSchema = type({
    chainId: ChainIdSchema,
    address: string(),
    symbol: string(),
    name: string(),
    decimals: number(),
    icon: optional(nullable(string())),
  });

  const EmptyObjectSchema = object({});

  const SrcChainStatusSchema = type({
    chainId: ChainIdSchema,
    txHash: optional(string()),
    amount: optional(string()),
    token: optional(union([EmptyObjectSchema, AssetSchema])),
  });

  const DestChainStatusSchema = type({
    chainId: ChainIdSchema,
    txHash: optional(string()),
    amount: optional(string()),
    token: optional(union([EmptyObjectSchema, AssetSchema])),
  });

  const RefuelStatusResponseSchema = object();

  const StatusResponseSchema = type({
    status: enums(Object.values(StatusTypes)),
    srcChain: SrcChainStatusSchema,
    destChain: optional(DestChainStatusSchema),
    bridge: optional(string()),
    isExpectedToken: optional(boolean()),
    isUnrecognizedRouterAddress: optional(boolean()),
    refuel: optional(RefuelStatusResponseSchema),
  });

  assert(data, StatusResponseSchema);
};
