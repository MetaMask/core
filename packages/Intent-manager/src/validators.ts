import { isValidHexAddress } from '@metamask/controller-utils';
import type { Infer } from '@metamask/superstruct';
import {
  string,
  boolean,
  number,
  type,
  optional,
  enums,
  define,
  union,
  pattern,
} from '@metamask/superstruct';
import { isStrictHexString } from '@metamask/utils';

// Helper schemas for Intent validation
const HexAddressSchema = define<string>('HexAddress', (v: unknown) =>
  isValidHexAddress(v as string, { allowNonPrefixed: false }),
);

const HexStringSchema = define<string>('HexString', (v: unknown) =>
  isStrictHexString(v as string),
);

const TruthyDigitStringSchema = pattern(string(), /^\d+$/u);

// Allow digit strings for amounts/validTo for flexibility across providers
const DigitStringOrNumberSchema = union([TruthyDigitStringSchema, number()]);

// Intent support (e.g., CoW Swap EIP-712 order signing)
const IntentProtocolSchema = enums(['cowswap']);

export const IntentOrderSchema = type({
  // EIP-712 Order fields (subset required for signing/submission)
  sellToken: HexAddressSchema,
  buyToken: HexAddressSchema,
  receiver: optional(HexAddressSchema),
  validTo: DigitStringOrNumberSchema,
  appData: string(),
  appDataHash: HexStringSchema,
  feeAmount: TruthyDigitStringSchema,
  kind: enums(['sell', 'buy']),
  partiallyFillable: boolean(),
  // One of these is required by CoW depending on kind; we keep both optional here and rely on backend validation
  sellAmount: optional(TruthyDigitStringSchema),
  buyAmount: optional(TruthyDigitStringSchema),
  // Optional owner/from for convenience when building domain/message
  from: optional(HexAddressSchema),
});

export const IntentSchema = type({
  protocol: IntentProtocolSchema,
  order: IntentOrderSchema,
  // Optional metadata to aid submission/routing
  settlementContract: optional(HexAddressSchema),
  relayer: optional(HexAddressSchema),
});

// Export types for use in other modules
export type IntentOrder = Infer<typeof IntentOrderSchema>;
export type Intent = Infer<typeof IntentSchema>;
