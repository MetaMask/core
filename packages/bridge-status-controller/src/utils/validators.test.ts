import { validateBridgeStatusResponse } from './validators';

const BridgeTxStatusResponses = {
  STATUS_PENDING_VALID: {
    status: 'PENDING',
    bridge: 'across',
    srcChain: {
      chainId: 42161,
      txHash:
        '0x76a65e4cea35d8732f0e3250faed00ba764ad5a0e7c51cb1bafbc9d76ac0b325',
      amount: '991250000000000',
      token: {
        address: '0x0000000000000000000000000000000000000000',
        assetId:
          'eip155:42161/erc20:0x82af49447d8a07e3bd95bd0d56f35241523fbab1',
        chainId: 42161,
        symbol: 'ETH',
        decimals: 18,
        name: 'ETH',
        coinKey: 'ETH',
        logoURI:
          'https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2/logo.png',
        priceUSD: '2550.12',
        icon: 'https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2/logo.png',
      },
    },
    destChain: {
      chainId: 10,
      token: {},
    },
  },
  STATUS_PENDING_VALID_MISSING_FIELDS: {
    status: 'PENDING',
    srcChain: {
      chainId: 42161,
      txHash:
        '0x5cbda572c686a5a57fe62735325e408f9164f77a4787df29ce13edef765adaa9',
    },
  },
  STATUS_PENDING_VALID_MISSING_FIELDS_2: {
    status: 'PENDING',
    bridge: 'hop',
    srcChain: {
      chainId: 42161,
      txHash:
        '0x5cbda572c686a5a57fe62735325e408f9164f77a4787df29ce13edef765adaa9',
      amount: '991250000000000',
      token: {
        chainId: 42161,
        assetId:
          'eip155:42161/erc20:0x82af49447d8a07e3bd95bd0d56f35241523fbab1',
        address: '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
        symbol: 'ETH',
        name: 'Ethereum',
        decimals: 18,
        icon: 'https://media.socket.tech/tokens/all/ETH',
        logoURI: 'https://media.socket.tech/tokens/all/ETH',
        chainAgnosticId: null,
      },
    },
  },
  STATUS_PENDING_INVALID_MISSING_FIELDS: {
    status: 'PENDING',
    bridge: 'across',
    srcChain: {
      chainId: 42161,
      txHash:
        '0x76a65e4cea35d8732f0e3250faed00ba764ad5a0e7c51cb1bafbc9d76ac0b325',
      amount: '991250000000000',
      token: {
        address: '0x0000000000000000000000000000000000000000',
        assetId:
          'eip155:42161/erc20:0x82af49447d8a07e3bd95bd0d56f35241523fbab1',
        chainId: 42161,
        symbol: 'ETH',
        decimals: 18,
        name: 'ETH',
        coinKey: 'ETH',
        logoURI:
          'https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2/logo.png',
        priceUSD: '2550.12',
        icon: 'https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2/logo.png',
      },
    },
    destChain: {
      token: {},
    },
  },
  STATUS_COMPLETE_VALID: {
    status: 'COMPLETE',
    isExpectedToken: true,
    bridge: 'across',
    srcChain: {
      chainId: 10,
      txHash:
        '0x9fdc426692aba1f81e145834602ed59ed331054e5b91a09a673cb12d4b4f6a33',
      amount: '4956250000000000',
      token: {
        address: '0x0000000000000000000000000000000000000000',
        assetId: 'eip155:10/erc20:0x4200000000000000000000000000000000000006',
        chainId: 10,
        symbol: 'ETH',
        decimals: 18,
        name: 'ETH',
        coinKey: 'ETH',
        logoURI:
          'https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2/logo.png',
        priceUSD: '2649.21',
        icon: 'https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2/logo.png',
      },
    },
    destChain: {
      chainId: 42161,
      txHash:
        '0x3a494e672717f9b1f2b64a48a19985842d82d0747400fccebebc7a4e99c8eaab',
      amount: '4926701727965948',
      token: {
        address: '0x0000000000000000000000000000000000000000',
        assetId: 'eip155:8453/erc20:0x4200000000000000000000000000000000000006',
        chainId: 42161,
        symbol: 'ETH',
        decimals: 18,
        name: 'ETH',
        coinKey: 'ETH',
        logoURI:
          'https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2/logo.png',
        priceUSD: '2648.72',
        icon: 'https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2/logo.png',
      },
    },
  },
  STATUS_COMPLETE_VALID_MISSING_FIELDS: {
    status: 'COMPLETE',
    bridge: 'across',
    srcChain: {
      chainId: 10,
      txHash:
        '0x9fdc426692aba1f81e145834602ed59ed331054e5b91a09a673cb12d4b4f6a33',
      amount: '4956250000000000',
      token: {
        address: '0x0000000000000000000000000000000000000000',
        assetId: 'eip155:10/erc20:0x4200000000000000000000000000000000000006',
        chainId: 10,
        symbol: 'ETH',
        decimals: 18,
        name: 'ETH',
        coinKey: 'ETH',
        logoURI:
          'https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2/logo.png',
        priceUSD: '2649.21',
        icon: 'https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2/logo.png',
      },
    },
    destChain: {
      chainId: 42161,
      txHash:
        '0x3a494e672717f9b1f2b64a48a19985842d82d0747400fccebebc7a4e99c8eaab',
      amount: '4926701727965948',
      token: {
        assetId: 'eip155:8453/erc20:0x4200000000000000000000000000000000000006',
        address: '0x0000000000000000000000000000000000000000',
        chainId: 42161,
        symbol: 'ETH',
        decimals: 18,
        name: 'ETH',
        coinKey: 'ETH',
        logoURI:
          'https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2/logo.png',
        priceUSD: '2648.72',
        icon: 'https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2/logo.png',
      },
    },
  },
  STATUS_COMPLETE_VALID_MISSING_FIELDS_2: {
    status: 'COMPLETE',
    isExpectedToken: false,
    bridge: 'across',
    srcChain: {
      chainId: 10,
      txHash:
        '0x4c57876fad21fb5149af5a58a4aba2ca9d6b212014505dd733b75667ca4f0f2b',
      amount: '991250000000000',
      token: {
        chainId: 10,
        assetId: 'eip155:10/erc20:0x4200000000000000000000000000000000000006',
        address: '0x4200000000000000000000000000000000000006',
        symbol: 'WETH',
        name: 'Wrapped Ether',
        decimals: 18,
        icon: 'https://media.socket.tech/tokens/all/WETH',
        // logoURI: 'https://media.socket.tech/tokens/all/WETH',
        // chainAgnosticId: 'ETH',
      },
    },
    destChain: {
      chainId: 8453,
      txHash:
        '0x60c4cad7c3eb14c7b3ace40cd4015b90927dadacbdc8673f404bea6a5603844b',
      amount: '988339336750062',
      token: {
        chainId: 8453,
        assetId: 'eip155:8453/erc20:0x4200000000000000000000000000000000000006',
        address: '0x4200000000000000000000000000000000000006',
        symbol: 'WETH',
        name: 'Wrapped Ether',
        decimals: 18,
        icon: null,
        // logoURI: null,
        // chainAgnosticId: null,
      },
    },
  },
  STATUS_COMPLETE_INVALID_MISSING_FIELDS: {
    status: 'COMPLETE',
    isExpectedToken: true,
    bridge: 'across',
  },
  STATUS_FAILED_VALID: {
    status: 'FAILED',
    bridge: 'across',
    srcChain: {
      chainId: 42161,
      txHash:
        '0x4c57876fad21fb5149af5a58a4aba2ca9d6b212014505dd733b75667ca4f0f2b',
      token: {},
    },
  },
  STATUS_SQUID_VALID: {
    status: 'COMPLETE',
    isExpectedToken: true,
    bridge: 'axelar',
    srcChain: {
      chainId: 10,
      txHash:
        '0x9fdc426692aba1f81e145834602ed59ed331054e5b91a09a673cb12d4b4f6a33',
    },
    destChain: {
      chainId: 42161,
      txHash:
        '0x3a494e672717f9b1f2b64a48a19985842d82d0747400fccebebc7a4e99c8eaab',
    },
  },
};

describe('validators', () => {
  describe('bridgeStatusValidator', () => {
    it.each([
      {
        input: BridgeTxStatusResponses.STATUS_PENDING_VALID,
        description: 'valid pending bridge status',
      },
      {
        input: BridgeTxStatusResponses.STATUS_PENDING_VALID_MISSING_FIELDS,
        description: 'valid pending bridge status missing fields',
      },
      {
        input: BridgeTxStatusResponses.STATUS_PENDING_VALID_MISSING_FIELDS_2,
        description: 'valid pending bridge status missing fields 2',
      },
      {
        input: BridgeTxStatusResponses.STATUS_COMPLETE_VALID,
        description: 'valid complete bridge status',
      },
      {
        input: BridgeTxStatusResponses.STATUS_COMPLETE_VALID_MISSING_FIELDS_2,
        description: 'complete bridge status with missing fields 2',
      },
      {
        input: BridgeTxStatusResponses.STATUS_COMPLETE_VALID_MISSING_FIELDS,
        description: 'complete bridge status with missing fields',
      },
      {
        input: BridgeTxStatusResponses.STATUS_FAILED_VALID,
        description: 'valid failed bridge status',
      },
      {
        input: BridgeTxStatusResponses.STATUS_SQUID_VALID,
        description: 'valid squid bridge status',
      },
      {
        input: {
          status: 'COMPLETE',
          srcChain: {
            chainId: 1151111081099710,
            txHash:
              '33LfknAQsrLC1WzmNybkZWUtuGANRFHNupsQ1YLCnjXGXxbBE93BbVTeKLLdE7Sz3WUdxnFW5HQhPuUayrXyqWky',
          },
        },
        description: 'placeholder complete swap status',
      },
    ])(
      'should not throw for valid response for $description',
      ({ input }: { input: unknown }) => {
        expect(() => validateBridgeStatusResponse(input)).not.toThrow();
      },
    );

    it.each([
      {
        input: BridgeTxStatusResponses.STATUS_PENDING_INVALID_MISSING_FIELDS,
        description: 'pending bridge status with missing fields',
      },
      {
        input: BridgeTxStatusResponses.STATUS_COMPLETE_INVALID_MISSING_FIELDS,
        description: 'complete bridge status with missing fields',
      },
      {
        input: undefined,
        description: 'undefined',
      },
      {
        input: null,
        description: 'null',
      },
      {
        description: 'empty object',
        input: {},
      },
    ])(
      'should throw for invalid response for $description',
      ({ input }: { input: unknown }) => {
        const mockConsoleError = jest
          .spyOn(console, 'error')
          .mockImplementation((_message: string) => jest.fn());

        // eslint-disable-next-line jest/require-to-throw-message
        expect(() => validateBridgeStatusResponse(input)).toThrow();
        // eslint-disable-next-line jest/no-restricted-matchers
        expect(mockConsoleError.mock.calls).toMatchSnapshot();
      },
    );
  });
});
