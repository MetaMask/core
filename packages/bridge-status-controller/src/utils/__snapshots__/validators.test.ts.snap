// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validators bridgeStatusValidator should throw for invalid response for complete bridge status with missing fields 1`] = `
Array [
  Array [
    "Bridge status validation failed",
    Object {
      "srcChain": "[across] Expected an object, but received: undefined",
    },
  ],
]
`;

exports[`validators bridgeStatusValidator should throw for invalid response for empty object 1`] = `
Array [
  Array [
    "Bridge status validation failed",
    Object {
      "srcChain": "[unknown] Expected an object, but received: undefined",
      "status": "[unknown] Expected one of \`\\"UNKNOWN\\",\\"FAILED\\",\\"PENDING\\",\\"COMPLETE\\"\`, but received: undefined",
    },
  ],
]
`;

exports[`validators bridgeStatusValidator should throw for invalid response for null 1`] = `
Array [
  Array [
    "Bridge status validation failed",
    Object {
      "unknown": "[unknown] Expected an object, but received: null",
    },
  ],
]
`;

exports[`validators bridgeStatusValidator should throw for invalid response for pending bridge status with missing fields 1`] = `
Array [
  Array [
    "Bridge status validation failed",
    Object {
      "destChain.chainId": "[across] Expected a string, but received: undefined",
    },
  ],
]
`;

exports[`validators bridgeStatusValidator should throw for invalid response for undefined 1`] = `
Array [
  Array [
    "Bridge status validation failed",
    Object {
      "unknown": "[unknown] Expected an object, but received: undefined",
    },
  ],
]
`;
